North-Seattle-College/ad440-winter2021-thursday-repo
# Natalia issue6 temp

Note: this PR repurposed a branch on a forked repo where all commit dates are retained.

Sprint1 [Task4](https://github.com/North-Seattle-College/ad440-winter2021-thursday-repo/issues/6)
This feature implements the API endpoint for getting, creating, updating and deleting a task when a userId and a taskId are passed in by the client.

Testing instructions:
* `preview` navigate to [nsc-functionsapp-team1](https://portal.azure.com/#@nsccloud.onmicrosoft.com/resource/subscriptions/9f4dcf43-aa06-457b-b975-f0216baef20d/resourceGroups/nsc-functions-team1/providers/Microsoft.Web/sites/nsc-functionsapp-team1/appServices)
*  **there have been issues with the group deployment, while we are fixing them, navigate [here](https://portal.azure.com/#@nsccloud.onmicrosoft.com/resource/subscriptions/9f4dcf43-aa06-457b-b975-f0216baef20d/resourceGroups/nataliasprint1test/providers/Microsoft.Web/sites/NataliaSprint1Test/appServices) and start the function if it happens to be stopped, then proceed**
* select "functions" on the left hand side blade
* click on `HttpTriggerAPIUsersIdTaskId` function
* select "Code + Test" on the left hand side blade - the code appears
* select Test/Run in the middle of the top-side menu - a parameters window appears
* Select the appropriate HTTP method
* Click on "+Add parameter" two times
* Enter two parameters {names} and {values} like so: 
userId<name>  1<value>
taskId<name> 18<value>
*  Button "RUN" lights up - click on it
* Results are displayed in the Output Window
Note: GET and UDPATE are further along than POST and DELETE. You will therefore see different response messages

GET input/output
<img width="1376" alt="Screen Shot 2021-01-20 at 5 14 45 PM" src="https://user-images.githubusercontent.com/46657216/105269864-9a09b600-5b49-11eb-8abb-2fea109a3418.png">
<img width="1374" alt="Screen Shot 2021-01-20 at 5 14 56 PM" src="https://user-images.githubusercontent.com/46657216/105269871-9c6c1000-5b49-11eb-8a29-90dc48e565bf.png">
UPDATE input/output
<img width="1377" alt="Screen Shot 2021-01-20 at 5 15 16 PM" src="https://user-images.githubusercontent.com/46657216/105269872-9d9d3d00-5b49-11eb-82c3-c3b128b1045c.png">
<img width="1374" alt="Screen Shot 2021-01-20 at 5 15 42 PM" src="https://user-images.githubusercontent.com/46657216/105269873-9e35d380-5b49-11eb-9241-d282a6b03470.png">
POST input/output
<img width="1378" alt="Screen Shot 2021-01-20 at 5 15 55 PM" src="https://user-images.githubusercontent.com/46657216/105269874-9e35d380-5b49-11eb-9ecc-ba4b82efdf80.png">
<img width="1374" alt="Screen Shot 2021-01-20 at 5 16 07 PM" src="https://user-images.githubusercontent.com/46657216/105269877-9ece6a00-5b49-11eb-88d8-998f95080251.png">
DELETE input/output
<img width="1383" alt="Screen Shot 2021-01-20 at 5 16 33 PM" src="https://user-images.githubusercontent.com/46657216/105269879-9ece6a00-5b49-11eb-811b-d6a7736a5e13.png">
<img width="1372" alt="Screen Shot 2021-01-20 at 5 16 45 PM" src="https://user-images.githubusercontent.com/46657216/105269881-9f670080-5b49-11eb-9a45-7f477bf3e440.png">

Steps taken to complete this task:
Date | Activity | Time Spent
------------ | ------------- | -------------
Jan 8-10 | AZURE odbc connection review | 4 hours
Jan 10-18 | Azure Functions in Python documentation review | 6 hours
Jan 9 | Setting up the local environment for testing | 4 hours
Jan 9-10 | Deployment of own test rg group, database, function app | 3 hours
Jan 13-20 | Development/debugging/testing/deployment | 10 hours
Jan17-20 | Team check-ins, documentation creation, git work, other misc |  4 hours
 |  | Total: 27 hours

**If you choose to test the function code on your local machine /deploy to cloud:**
ensure you have access to Azure
clone the repository
*ensure that code for the other API functions is present as there are some shared modules across the functions * 
install python 3.6-3.8*
- install required libraries (azure.functions, pyodbc)
- install Azure Functions Core Tools and in Visual Studio Code
- install Python and Visual Studio Code with the Azure Function extension
-set up the environ (venv and vscode files are not part of the available code)
-update `yourlocal.settings.jason` file to include these values, which will permit connecting to the database deployed to the cloud:
{
  "IsEncrypted": false,
  "Values": {
    "AzureWebJobsStorage": "",
    "FUNCTIONS_WORKER_RUNTIME": "python",
    "ENV_DATABASE_USERNAME": "username",
    "ENV_DATABASE_PASSWORD": "xyz",
    "ENV_DATABASE_NAME":  "xyz",
    "ENV_DATABASE_SERVER": "xyz"
  }
}

-run your code with the appropriate venv Python interpreter

when testing in Postman, input same params as recommended above (userId =1, taskId =18)
*note that some values such as userId =1, taskId =1 are not available in the database*

To deploy in the cloud, create a resource group, a functions app, add records in the functions app configuration file to contain 
same information as above, for 
`"ENV_DATABASE_SERVER"`
`"ENV_DATABASE_NAME"`
`"ENV_DATABASE_USERNAME"`
`"ENV_DATABASE_PASSWORD"`
Deploy the function directly from VSCode via the Azure extension

For additional troubleshooting, see documentation for other API endpoints within the project, e.g. [here](https://github.com/North-Seattle-College/ad440-winter2021-thursday-repo/pull/49#pullrequestreview-572944842)
